ffi/src/engine_funcs.rs

```diff
@@ -0,0 +1,131 @@
+    }
+}
+
+// Intentionally opaque to the engine.
+pub struct FileReadResultIterator {
+    // Box -> Wrap its unsized content this struct is fixed-size with thin pointers.
+    // Item = Box<dyn EngineData>, see above, Vec<bool> -> can become a KernelBoolSlice
+    data: FileDataReadResultIterator,
+
+    // Also keep a reference to the external engine for its error allocator.
+    // Parquet and Json handlers don't hold any reference to the tokio reactor, so the iterator
+    // terminates early if the last engine goes out of scope.
+    engine: Arc<dyn ExternEngine>,
+}
+
+#[handle_descriptor(target=FileReadResultIterator, mutable=true, sized=true)]
+pub struct MutFileReadResultIterator;
+
+impl Drop for FileReadResultIterator {
+    fn drop(&mut self) {
+        debug!("dropping FileReadResultIterator");
+    }
+}
+
+/// Call the engine back with the next `EngingeData` batch read by Parquet/Json handler
+/// # Safety
+///
+/// The iterator must be valid (returned by [`read_parquet_files`]) and not yet freed by
+/// [`free_read_result_iter`]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn read_result_next(
+    mut data: Handle<MutFileReadResultIterator>,
+    engine_context: NullableCvoid,
+    engine_visitor: extern "C" fn(engine_context: NullableCvoid, engine_data: Handle<EngineData>),
+) -> ExternResult<bool> {
+    let iter = unsafe { data.as_mut() };
+    read_result_next_impl(iter, engine_context, engine_visitor)
+        .into_extern_result(iter.engine.error_allocator())
+}
+
+fn read_result_next_impl(
+    iter: &mut FileReadResultIterator,
+    engine_context: NullableCvoid,
+    engine_visitor: extern "C" fn(engine_context: NullableCvoid, engine_data: Handle<EngineData>),
+) -> DeltaResult<bool> {
+    if let Some(data) = iter.data.next().transpose()? {
+        (engine_visitor)(engine_context, data.into());
+        // TODO: calling `into_raw` in the visitor causes this to segfault
+        //       perhaps the callback needs to indicate if it took ownership or not
+        // unsafe { BoxHandle::drop_handle(data_handle) };
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// Free the memory from the passed read result iterator
+/// # Safety
+///
+/// Caller is responsible for (at most once) passing a valid pointer returned by a call to
+/// [`read_parquet_files`].
+#[no_mangle]
+pub unsafe extern "C" fn free_read_result_iter(data: Handle<MutFileReadResultIterator>) {
+    data.drop_handle();
+}
+
+/// Use the specified engine's [`delta_kernel::ParquetHandler`] to read the specified file.
+///
+/// # Safety
+/// Caller is responsible for calling with a valid `ExternEngineHandle` and `FileMeta`
+#[no_mangle]
+pub unsafe extern "C" fn read_parquet_files(
+    engine: Handle<SharedExternEngine>,
+    file: &FileMeta,
+    physical_schema: Handle<SharedSchema>,
+) -> ExternResult<Handle<MutFileReadResultIterator>> {
+    let extern_engine = unsafe { engine.clone_as_arc() };
```

Comment:
This one I made the same. Below we need two names, because we use `engine` in some places, but at the end we want to put the `extern_engine` into the struct.